In [103]:
#Importing required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn
%matplotlib inline

In [104]:
#To ignore warnings
import warnings
warnings.filterwarnings("ignore")

In [105]:
#Read the csv file and putting it inot 'df' object.
df=pd.read_csv('adult.csv')

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              48842 non-null  int64 
 1   workclass        48842 non-null  object
 2   fnlwgt           48842 non-null  int64 
 3   education        48842 non-null  object
 4   educational-num  48842 non-null  int64 
 5   marital-status   48842 non-null  object
 6   occupation       48842 non-null  object
 7   relationship     48842 non-null  object
 8   race             48842 non-null  object
 9   gender           48842 non-null  object
 10  capital-gain     48842 non-null  int64 
 11  capital-loss     48842 non-null  int64 
 12  hours-per-week   48842 non-null  int64 
 13  native-country   48842 non-null  object
 14  income           48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [107]:
# Lets understand how data look
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

You can observe that the columns workclass and occupation consist of missing value which are represented as '?' in the dataframe.

On looking a bit more closely, you will also find that whenever workclass is having a missing value, occupation is also missing in that row. Let's check how may rows are missing.





In [108]:
#rows with missing values represented as '?' .
df_1=df[df.workclass == '?']
df_1

age workclass  fnlwgt     education  educational-num  \
4       18         ?  103497  Some-college               10   
6       29         ?  227026       HS-grad                9   
13      58         ?  299831       HS-grad                9   
22      72         ?  132015       7th-8th                4   
35      65         ?  191846       HS-grad                9   
...    ...       ...     ...           ...              ...   
48811   35         ?  320084     Bachelors               13   
48812   30         ?   33811     Bachelors               13   
48820   71         ?  287372     Doctorate               16   
48822   41         ?  202822       HS-grad                9   
48823   72         ?  129912       HS-grad                9   

           marital-status occupation   relationship                race  \
4           Never-married          ?      Own-child               White   
6           Never-married          ?      Unmarried               Black   
13     Married-civ-spouse          ?        Husband               White   
22               Divorced          ?  Not-in-family               White   
35     Married-civ-spouse          ?        Husband               White   
...                   ...        ...            ...                 ...   
48811  Married-civ-spouse          ?           Wife               White   
48812       Never-married          ?  Not-in-family  Asian-Pac-Islander   
48820  Married-civ-spouse          ?        Husband               White   
48822           Separated          ?  Not-in-family               Black   
48823  Married-civ-spouse          ?        Husband               White   

       gender  capital-gain  capital-loss  hours-per-week native-country  \
4      Female             0             0              30  United-States   
6        Male             0             0              40  United-States   
13       Male             0             0              35  United-States   
22     Female             0             0               6  United-States   
35       Male             0             0              40  United-States   
...       ...           ...           ...             ...            ...   
48811  Female             0             0              55  United-States   
48812  Female             0             0              99  United-States   
48820    Male             0             0              10  United-States   
48822  Female             0             0              32  United-States   
48823    Male             0             0              25  United-States   

      income  
4      <=50K  
6      <=50K  
13     <=50K  
22     <=50K  
35     <=50K  
...      ...  
48811   >50K  
48812  <=50K  
48820   >50K  
48822  <=50K  
48823  <=50K  

[2799 rows x 15 columns]

Now we can check the number of rows in df_1.

In [109]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2799 entries, 4 to 48823
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              2799 non-null   int64 
 1   workclass        2799 non-null   object
 2   fnlwgt           2799 non-null   int64 
 3   education        2799 non-null   object
 4   educational-num  2799 non-null   int64 
 5   marital-status   2799 non-null   object
 6   occupation       2799 non-null   object
 7   relationship     2799 non-null   object
 8   race             2799 non-null   object
 9   gender           2799 non-null   object
 10  capital-gain     2799 non-null   int64 
 11  capital-loss     2799 non-null   int64 
 12  hours-per-week   2799 non-null   int64 
 13  native-country   2799 non-null   object
 14  income           2799 non-null   object
dtypes: int64(6), object(9)
memory usage: 349.9+ KB


There are 2799 rows with missing values , which is about 5% of the total data.We choose to simple drop these rows.

In [110]:
#dropping the rows having missing values in workclass.
df=df[df['workclass'] !='?']
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
5   34    Private  198693          10th                6       Never-married   

          occupation   relationship   race gender  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black   Male             0             0   
1    Farming-fishing        Husband  White   Male             0             0   
2    Protective-serv        Husband  White   Male             0             0   
3  Machine-op-inspct        Husband  Black   Male          7688             0   
5      Other-service  Not-in-family  White   Male             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
5              30  United-States  <=50K

Let's see whether any other columns contain a "?". Since '?" is a string , we can apply this check only on the categorical columns.

In [111]:
#select all categorical variables
df_categorical=df.select_dtypes(include=['object'])

#checking whether any other columns contain a "?"
df_categorical.apply(lambda x: x=="?" , axis=0).sum()

workclass           0
education           0
marital-status      0
occupation         10
relationship        0
race                0
gender              0
native-country    811
income              0
dtype: int64

Thus , the columns occupation  and native-country contains some "?"s.Let's get rid of them

In [112]:
#dropping the "?"s
df=df[df['occupation'] != '?']
df=df[df['native-country'] != '?']


Now we have a clean dataframe which is ready for model building

In [113]:
#clean dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45222 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              45222 non-null  int64 
 1   workclass        45222 non-null  object
 2   fnlwgt           45222 non-null  int64 
 3   education        45222 non-null  object
 4   educational-num  45222 non-null  int64 
 5   marital-status   45222 non-null  object
 6   occupation       45222 non-null  object
 7   relationship     45222 non-null  object
 8   race             45222 non-null  object
 9   gender           45222 non-null  object
 10  capital-gain     45222 non-null  int64 
 11  capital-loss     45222 non-null  int64 
 12  hours-per-week   45222 non-null  int64 
 13  native-country   45222 non-null  object
 14  income           45222 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.5+ MB


````

## Data preparation

There are a number of preprocessing steps we need to do before building the model.

Firstly, note that we have both categorical and numeric features as predictors. In previous model such as linear and logistic regression, we had created  dummy variables  for categorical variables, since these models (being mathematical
equations) can process only numeric variables. 

All that is not required in decision trees, since they can process categorical variables easily. However, we still need to encode the categorical variables into a standard format so that sklearn can understand them and build the tree. We'll do that using the LabelEncoder() class, which comes with sklearn.preprocessing

You can read the documentation of LabelEncoder.


In [114]:
from sklearn import preprocessing

#encode categorical variables using label Encoder

#select all categorical variables
df_categorical = df.select_dtypes(include=['object'])
df_categorical.head()

workclass     education      marital-status         occupation  \
0    Private          11th       Never-married  Machine-op-inspct   
1    Private       HS-grad  Married-civ-spouse    Farming-fishing   
2  Local-gov    Assoc-acdm  Married-civ-spouse    Protective-serv   
3    Private  Some-college  Married-civ-spouse  Machine-op-inspct   
5    Private          10th       Never-married      Other-service   

    relationship   race gender native-country income  
0      Own-child  Black   Male  United-States  <=50K  
1        Husband  White   Male  United-States  <=50K  
2        Husband  White   Male  United-States   >50K  
3        Husband  Black   Male  United-States   >50K  
5  Not-in-family  White   Male  United-States  <=50K

In [115]:
# apply Label encoder to df_categorical

le=preprocessing.LabelEncoder()
df_categorical=df_categorical.apply(le.fit_transform)
df_categorical.head()

workclass  education  marital-status  occupation  relationship  race  \
0          2          1               4           6             3     2   
1          2         11               2           4             0     4   
2          1          7               2          10             0     4   
3          2         15               2           6             0     2   
5          2          0               4           7             1     4   

   gender  native-country  income  
0       1              38       0  
1       1              38       0  
2       1              38       1  
3       1              38       1  
5       1              38       0

In [116]:
# concat df_categorical with original df

df=df.drop(df_categorical, axis=1)
df=pd.concat([df , df_categorical], axis=1)
df.head()

age  fnlwgt  educational-num  capital-gain  capital-loss  hours-per-week  \
0   25  226802                7             0             0              40   
1   38   89814                9             0             0              50   
2   28  336951               12             0             0              40   
3   44  160323               10          7688             0              40   
5   34  198693                6             0             0              30   

   workclass  education  marital-status  occupation  relationship  race  \
0          2          1               4           6             3     2   
1          2         11               2           4             0     4   
2          1          7               2          10             0     4   
3          2         15               2           6             0     2   
5          2          0               4           7             1     4   

   gender  native-country  income  
0       1              38       0  
1       1              38       0  
2       1              38       1  
3       1              38       1  
5       1              38       0

In [117]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45222 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   age              45222 non-null  int64
 1   fnlwgt           45222 non-null  int64
 2   educational-num  45222 non-null  int64
 3   capital-gain     45222 non-null  int64
 4   capital-loss     45222 non-null  int64
 5   hours-per-week   45222 non-null  int64
 6   workclass        45222 non-null  int32
 7   education        45222 non-null  int32
 8   marital-status   45222 non-null  int32
 9   occupation       45222 non-null  int32
 10  relationship     45222 non-null  int32
 11  race             45222 non-null  int32
 12  gender           45222 non-null  int32
 13  native-country   45222 non-null  int32
 14  income           45222 non-null  int32
dtypes: int32(9), int64(6)
memory usage: 4.0 MB


In [118]:
#convert target variable income to categorical
df['income'] = df['income'].astype('category')

Now all the categorical variables are suitably encoded . Let's build the model

````

## Model Building and Evaulation

Let's first build a decsision tree with default hyperparamters. Then we'll use cross-validation to tune them.

In [119]:
#Import train-train split
from sklearn.model_selection import train_test_split

In [120]:
#Putting features variable to X
X=df.drop('income' , axis=1)

#Putting response variable to y
y=df['income']

In [121]:
# Splitting the data into train and test
X_train , X_test , y_train , y_test = train_test_split(X , y , 
                                            test_size=0.30 , 
                                            random_state=99)
X_train.head()

age  fnlwgt  educational-num  capital-gain  capital-loss  \
39923   50   45961               13          6849             0   
19651   38  348739               15             0             0   
11260   29  198210                9             0             0   
28311   45  217654               10             0             0   
48075   30  204470               10             0             0   

       hours-per-week  workclass  education  marital-status  occupation  \
39923              40          5          9               3           9   
19651              60          2         14               2           9   
11260              40          2         11               4          12   
28311              40          2         15               2           5   
48075              36          3         15               2           3   

       relationship  race  gender  native-country  
39923             1     4       1              38  
19651             0     4       1              38  
11260             4     2       0              38  
28311             0     4       1              10  
48075             0     4       1              38

In [136]:
# Importing decision tree classifier from sklearn library

from sklearn.tree import DecisionTreeClassifier

#Fitting the decision tree with default hyperparameters, apart from
#  #max_depth which is 5 so that we can plot and read the tree. 
dt_default = DecisionTreeClassifier(max_depth=5)
dt_default.fit(X_train , y_train)


DecisionTreeClassifier(max_depth=5)

In [137]:
# Let's check the evaluation metrics of our default model 

# Importing classification report and confusion matrix from sklearn metrics 
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Making predictions
y_pred_default = dt_default.predict(X_test)

#Printing classification report 
print(classification_report(y_test, y_pred_default))





              precision    recall  f1-score   support

           0       0.87      0.94      0.90     10223
           1       0.75      0.56      0.64      3344

    accuracy                           0.85     13567
   macro avg       0.81      0.75      0.77     13567
weighted avg       0.84      0.85      0.84     13567



In [138]:
# Printing confusion matrix and accuracy
print(confusion_matrix(y_test, y_pred_default))
print(accuracy_score(y_test, y_pred_default))

[[9611  612]
 [1469 1875]]
0.8466131053291074


```



## PLotting the Decision Tree

To visualize deccision tree in python , you need to install certain external lobraries.

In [139]:
from IPython.display import Image 
from sklearn.tree import export_graphviz 
import pydotplus
import graphviz

# Putting features
features = list(df.columns[1:])
features


['fnlwgt',
 'educational-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'workclass',
 'education',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'gender',
 'native-country',
 'income']

In [126]:


dot_data = StringIO()
export_graphviz(dt_default, out_file=dot_data,
                feature_names=features, filled=True, rounded=True)

graph = pydotplus.graph_from_dot_data(dot_data.getvalue())

Image(graph.create_png())


InvocationException: GraphViz's executables not found

````

### Hyperparameter Tuning

The default tree is quite complex, and we need to simplify it by tuning the hyperparameters.

First, let's understand the parameters in a decision tree. You can read this in the documentation using help (DecisionTreeClassifier).

⚫ criterion (Gini/IG or entropy): It defines the function to measure the quality of a split. Sklearn supports "gini" criteria for Gini Index & "entropy" for Information Gain. By default, it takes the value "gini".

⚫ splitter: It defines the strategy to choose the split at each node. Supports "best" value to choose the best split & "random" to choose the best random split. By default, it takes "best" value.

⚫ max features: It defines the no. of features to consider when looking for the best split. We can input integer, float, string & None value.

If an integer is inputted then it considers that value as max features at each split.

If float value is taken then it shows the percentage of features at each split.

If "auto" or "sqrt" is taken then max features=sqrt(n_features).

If "log2" is taken then max features = log2(n_features).

If None, then max features = n_features. By default, it takes "None" value.

⚫max_depth: The max_depth parameter denotes maximum depth of the tree. It can take any integer value or None. If None, then nodes are expanded until all leaves are pure or until all leaves contain less than min_samples_split samples. By default, it takes "None" value.

⚫min samples split: This tells above the minimum no, of samples internal node. If an integer value is

taken then consider min samples split as the minimum no. If float, the rows percentage. By default, it takes "2" value.

⚫ min samples leat The minimum number of samples required to be at a leaf node. If an integer value is taken then

consider --min samples leaf as the minimum no. It float, then it shows percentage. By default, it takes "1" value

⚫ max leaf nodes: it defines the maximum number of possible leaf nodes. It None then it takes an unlimited number of

leaf nodes. By default, it takes "None" value.

⚫ min impurity split it defines the threshold for early stopping tree growth. A node will split if its impunity is above the threshold otherwise it is a leaf.


```

## Tuning max_depth

Let's first try to find the optimum values for max_depth and understand how the value of max_depth affects the decision tree

Here, we are creating a dataframe with max_depth in range 1 to 80 and checking the accuracy score corresponding to each max_depth.


To reiterate, a grid search scheme consists of:

- an estimator (classifier such as SVC() or decisico tree)

- a parameter space

- a method for searching or sampling candidates (optional)

- a cross-validation scheme, and

* score function (accuracy, roc_auc etc.)

In [127]:
 #GridSearchCV to find optimal max_depth
from sklearn. model_selection import KFold 
from sklearn.model_selection import GridSearchCV


#specify number of folds for k-fold CV
n_folds = 5

# parameters to build the model on
parameters = {"max_depth": range(1, 40)}

# instantiate the model
dtree = DecisionTreeClassifier (criterion = "gini", 
                                random_state= 100)

# fit tree on training data 
tree = GridSearchCV(dtree, parameters,cv=n_folds,
                     scoring="accuracy")

tree.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=100),
             param_grid={'max_depth': range(1, 40)}, scoring='accuracy')

In [128]:
#scores of GridSearch CV
scores = tree.cv_results_
pd.DataFrame(scores).head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.025479      0.002998         0.026942        0.043823   
1       0.022443      0.004604         0.003125        0.001058   
2       0.030560      0.003629         0.003412        0.000926   
3       0.038568      0.004059         0.003352        0.001541   
4       0.060097      0.014537         0.003027        0.001410   

  param_max_depth            params  split0_test_score  split1_test_score  \
0               1  {'max_depth': 1}           0.751698           0.751540   
1               2  {'max_depth': 2}           0.824988           0.819460   
2               3  {'max_depth': 3}           0.835413           0.833360   
3               4  {'max_depth': 4}           0.844732           0.840310   
4               5  {'max_depth': 5}           0.850419           0.847102   

   split2_test_score  split3_test_score  split4_test_score  mean_test_score  \
0           0.751540           0.751540           0.751540         0.751572   
1           0.821671           0.821987           0.820250         0.821671   
2           0.834781           0.834781           0.833518         0.834371   
3           0.843785           0.842995           0.838888         0.842142   
4           0.847102           0.848365           0.846944         0.847986   

   std_test_score  rank_test_score  
0        0.000063               39  
1        0.001898               17  
2        0.000797               13  
3        0.002195               10  
4        0.001320                6

Now let's visualize how train and test score changes with max_depth.

In [130]:

# plotting accuracies with max_depth 
plt.figure()
plt.plot(scores["param_max_depth"], 
         scores["mean_train_score"], 
         label="training accuracy")

plt.plot(scores["param_max_depth"], 
         scores ["mean_test_score"],
         label="test accuracy") 
         
plt.xlabel("max_depth")

plt.ylabel("Accuracy")

plt.legend()
plt.show()

KeyError: 'mean_train_score'

<Figure size 640x480 with 0 Axes>

You can see that as we increase the value of max_depth, both training and test score increase till about max-depth = 10, after which the test score gradually reduces. Note that the scores are average accuracies across the 5-folds.

Thus, it is clear that the model is overfitting the training data if the max depth is too high. Next, let's see how the model behaves with other hyperparameters.


````````

## Tuning min_samples_leaf

The hyperparameter min_samples_leaf indicates the minimum number of samples required to be at a leaf

So if the values of min_samples_leaf is less, say 5, then the will be constructed even if a leaf has 5, 6 etc, observations and is likely to overfit).

Let's see what will be the optimum Value for min samples leaf.


In [151]:
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Specify the number of folds for k-fold CV
n_folds = 5

# Parameters to build the model on
parameters = {'min_samples_leaf': range(5, 200, 20)}

# Instantiate the model
dtree = DecisionTreeClassifier(criterion="gini", random_state=100)

# Fit the tree on the training data
tree = GridSearchCV(dtree, parameters, cv=n_folds, scoring="accuracy")
tree.fit(X_train, y_train)


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=100),
             param_grid={'min_samples_leaf': range(5, 200, 20)},
             scoring='accuracy')

In [152]:
# Scores of GridSearch CV
scores = tree.cv_results_
pd.DataFrame(scores).head()


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.105940      0.006778         0.005188        0.002264   
1       0.115677      0.017550         0.003552        0.001400   
2       0.097060      0.010533         0.004090        0.000991   
3       0.083368      0.005909         0.003407        0.001032   
4       0.077097      0.007340         0.002816        0.000464   

  param_min_samples_leaf                    params  split0_test_score  \
0                      5   {'min_samples_leaf': 5}           0.832570   
1                     25  {'min_samples_leaf': 25}           0.846944   
2                     45  {'min_samples_leaf': 45}           0.847733   
3                     65  {'min_samples_leaf': 65}           0.850419   
4                     85  {'min_samples_leaf': 85}           0.847733   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.824198           0.828305           0.828463           0.824040   
1           0.851998           0.846944           0.845522           0.847102   
2           0.855947           0.847575           0.852156           0.848523   
3           0.853578           0.847891           0.854367           0.851998   
4           0.854051           0.846944           0.849945           0.853104   

   mean_test_score  std_test_score  rank_test_score  
0         0.827515        0.003167               10  
1         0.847702        0.002223                9  
2         0.850387        0.003239                2  
3         0.851651        0.002318                1  
4         0.850355        0.002825                3

In [148]:

# plotting accuracies with max_depth 
plt.figure()
plt.plot(scores["param_min_samples_leaf"], 
         scores["mean_train_score"], 
         label="training accuracy")

plt.plot(scores["param_min_samples_leaf"], 
         scores ["mean_test_score"],
         label="test accuracy") 
         
plt.xlabel("min_samples_leaf")

plt.ylabel("Accuracy")

plt.legend()
plt.show()





KeyError: 'mean_train_score'

<Figure size 640x480 with 0 Axes>

You can see that at low values of min_samples_leaf, the tree gets a bit overfitted. At values > 100, however, the model becomes more stable and the training and test accuracy start to converge.


````````````````


## Tuning min_samples_split

The hyperparameter min samples_split is the minimum no. of samples required to split an internal node. Its default value is 2, which means that even if a node is having 2 samples it can be furthur divided into leaf nodes.

In [153]:
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Specify the number of folds for k-fold CV
n_folds = 5

# Parameters to build the model on
parameters = {'min_samples_split': range(5, 200, 20)}

# Instantiate the model
dtree = DecisionTreeClassifier(criterion="gini", random_state=100)

# Fit the tree on the training data
tree = GridSearchCV(dtree, parameters, cv=n_folds, scoring="accuracy")
tree.fit(X_train, y_train)


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=100),
             param_grid={'min_samples_split': range(5, 200, 20)},
             scoring='accuracy')

In [154]:
# Scores of GridSearch CV
scores = tree.cv_results_
pd.DataFrame(scores).head()


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.096274      0.003028         0.003050        0.000042   
1       0.116093      0.010277         0.003904        0.000737   
2       0.111155      0.008325         0.004273        0.001723   
3       0.099517      0.003054         0.002965        0.000858   
4       0.098451      0.004588         0.003146        0.000699   

  param_min_samples_split                     params  split0_test_score  \
0                       5   {'min_samples_split': 5}           0.816775   
1                      25  {'min_samples_split': 25}           0.837308   
2                      45  {'min_samples_split': 45}           0.840941   
3                      65  {'min_samples_split': 65}           0.842679   
4                      85  {'min_samples_split': 85}           0.844574   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.805402           0.814563           0.808087           0.812510   
1           0.823409           0.826884           0.831780           0.828305   
2           0.833991           0.837466           0.835255           0.839046   
3           0.839204           0.844732           0.842837           0.839362   
4           0.844574           0.846944           0.844890           0.842679   

   mean_test_score  std_test_score  rank_test_score  
0         0.811467        0.004173               10  
1         0.829537        0.004723                9  
2         0.837340        0.002508                8  
3         0.841763        0.002150                7  
4         0.844732        0.001355                6

In [155]:

# plotting accuracies with max_depth 
plt.figure()
plt.plot(scores["param_min_samples_split"], 
         scores["mean_train_score"], 
         label="training accuracy")

plt.plot(scores["param_min_samples_split"], 
         scores ["mean_test_score"],
         label="test accuracy") 
         
plt.xlabel("min_samples_split")

plt.ylabel("Accuracy")

plt.legend()
plt.show()

KeyError: 'mean_train_score'

<Figure size 640x480 with 0 Axes>

This shows that you increase the min_samples_split, the tree overfits lesser since the models is less complex

```````````````````````````

## Grid Search to Find Optimal Hyperparameters

We can now use GridSearchCV to find multiple optimal hyperparameters together. Note that this time, we'll also specify the criterion (gini/entropy or IG).


In [160]:
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Create the parameter grid
param_grid = {
    'max_depth': range(5, 15, 5),
    'min_samples_leaf': range(50, 150, 50),
    'min_samples_split': range(50, 150, 50),
    'criterion': ["entropy", "gini"]
}

n_folds = 5

# Instantiate the grid search model
dtree = DecisionTreeClassifier()

grid_search = GridSearchCV(estimator=dtree, param_grid=param_grid,
                           cv=n_folds, verbose=1)

# Fit the grid search to the data
grid_search.fit(X_train, y_train)


Fitting 5 folds for each of 16 candidates, totalling 80 fits


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['entropy', 'gini'],
                         'max_depth': range(5, 15, 5),
                         'min_samples_leaf': range(50, 150, 50),
                         'min_samples_split': range(50, 150, 50)},
             verbose=1)

##### Running the model with best parameters obtained from grid search

In [162]:
cv_results=pd.DataFrame(grid_search.cv_results_)
cv_results

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.045274      0.004788         0.002819        0.000755   
1        0.044111      0.003018         0.002709        0.000609   
2        0.044196      0.003886         0.002976        0.001251   
3        0.055326      0.013758         0.002799        0.000746   
4        0.067513      0.003375         0.002614        0.000768   
5        0.072158      0.006118         0.003017        0.000663   
6        0.067228      0.007217         0.002401        0.000494   
7        0.064998      0.005197         0.002589        0.000779   
8        0.036262      0.001688         0.002468        0.000914   
9        0.043496      0.003774         0.003800        0.000978   
10       0.043099      0.004913         0.002788        0.000394   
11       0.039844      0.002319         0.003073        0.001761   
12       0.067970      0.010378         0.002219        0.000453   
13       0.070460      0.004112         0.003410        0.001005   
14       0.064068      0.005910         0.002454        0.000905   
15       0.057538      0.002568         0.002992        0.000900   

   param_criterion param_max_depth param_min_samples_leaf  \
0          entropy               5                     50   
1          entropy               5                     50   
2          entropy               5                    100   
3          entropy               5                    100   
4          entropy              10                     50   
5          entropy              10                     50   
6          entropy              10                    100   
7          entropy              10                    100   
8             gini               5                     50   
9             gini               5                     50   
10            gini               5                    100   
11            gini               5                    100   
12            gini              10                     50   
13            gini              10                     50   
14            gini              10                    100   
15            gini              10                    100   

   param_min_samples_split                                             params  \
0                       50  {'criterion': 'entropy', 'max_depth': 5, 'min_...   
1                      100  {'criterion': 'entropy', 'max_depth': 5, 'min_...   
2                       50  {'criterion': 'entropy', 'max_depth': 5, 'min_...   
3                      100  {'criterion': 'entropy', 'max_depth': 5, 'min_...   
4                       50  {'criterion': 'entropy', 'max_depth': 10, 'min...   
5                      100  {'criterion': 'entropy', 'max_depth': 10, 'min...   
6                       50  {'criterion': 'entropy', 'max_depth': 10, 'min...   
7                      100  {'criterion': 'entropy', 'max_depth': 10, 'min...   
8                       50  {'criterion': 'gini', 'max_depth': 5, 'min_sam...   
9                      100  {'criterion': 'gini', 'max_depth': 5, 'min_sam...   
10                      50  {'criterion': 'gini', 'max_depth': 5, 'min_sam...   
11                     100  {'criterion': 'gini', 'max_depth': 5, 'min_sam...   
12                      50  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   
13                     100  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   
14                      50  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   
15                     100  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   

    split0_test_score  split1_test_score  split2_test_score  \
0            0.842047           0.847102           0.846312   
1            0.842047           0.847102           0.846312   
2            0.842047           0.846628           0.846154   
3            0.842047           0.846628           0.846154   
4            0.846786           0.850261           0.848365   
5            0.846786           0.850261           0.

In [163]:
# Printing the optimal accuracy score and hyperparameters
print("best accuracy" , grid_search.best_score_)
print(grid_search.best_estimator_)

best accuracy 0.8513031116727217
DecisionTreeClassifier(max_depth=10, min_samples_leaf=50, min_samples_split=50)


````

#### Running the model with best parameters obtained from grid search

In [164]:
# model with optimal hyperparameters
clf_gini=DecisionTreeClassifier(criterion="gini" , random_state=100 ,
                                 max_depth=10, min_samples_leaf=50, 
                                 min_samples_split=50)

clf_gini.fit(X_train , y_train)

DecisionTreeClassifier(max_depth=10, min_samples_leaf=50, min_samples_split=50,
                       random_state=100)

In [165]:
#accuracy score
clf_gini.score(X_test , y_test)

0.8516989754551485

In [166]:
# plotting the tree
from IPython.display import Image
from sklearn.tree import export_graphviz
import pydotplus
import graphviz
from io import StringIO



dot_data = StringIO()
export_graphviz(clf_gini, out_file=dot_data,
                feature_names=features, filled=True, rounded=True)

graph = pydotplus.graph_from_dot_data(dot_data.getvalue())

Image(graph.create_png())



InvocationException: GraphViz's executables not found

You can see this tree is too complex to understand. Let's try reducing the max_depth and see how the tree looks.


In [167]:
# model with optimal hyperparameters
clf_gini=DecisionTreeClassifier(criterion="gini" , random_state=100 ,
                                 max_depth=3, min_samples_leaf=50, 
                                 min_samples_split=50)

clf_gini.fit(X_train , y_train)

#score
print(clf_gini.score(X_test , y_test))

0.8344512419842264


In [168]:
# classification metrices
from sklearn.metrics import classification_report , confusion_matrix
y_pred = clf_gini.predict(X_test)
print(classification_report(y_test , y_pred))

              precision    recall  f1-score   support

           0       0.85      0.94      0.90     10223
           1       0.75      0.50      0.60      3344

    accuracy                           0.83     13567
   macro avg       0.80      0.72      0.75     13567
weighted avg       0.83      0.83      0.82     13567



In [169]:
# confusion metrix
print(confusion_matrix(y_test , y_pred))

[[9660  563]
 [1683 1661]]
